lib/Dialect/FIRRTL/Transforms/BlackBoxMemory.cpp

```diff
@@ -226,12 +238,63 @@ static FModuleOp createWrapperModule(MemOp op,
          llvm::enumerate(memPortType.cast<BundleType>().getElements())) {
       auto fieldValue =
           builder.create<SubfieldOp>(op.getLoc(), memPort, field.index());
-      // Create the connection between module arguments and the external module,
-      // making sure that sinks are on the LHS
-      if (!field.value().isFlip)
-        builder.create<ConnectOp>(op.getLoc(), *extResultIt, fieldValue);
-      else
-        builder.create<ConnectOp>(op.getLoc(), fieldValue, *extResultIt);
+      auto fieldType = field.value().type.cast<FIRRTLType>();
+      bool isReadPort = field.value().isFlip;
+      // Flatten the vector, only if the elements of the vector are of simple
+      // ground type with known bitwidth.
+      if (fieldType.isa<FVectorType>() && flattenVector &&
+          fieldType.dyn_cast<FVectorType>()
+                  .getElementType()
+                  .getBitWidthOrSentinel() > 0) {
+        // For read ports extract bits from the memory read port and distribute
+        // them over the vector elements. Example for an 8 bit vector element,
+        // wrapper_readPort[0] = Mem_readPort[7:0];
+        // wrapper_readPort[1] = Mem_readPort[15:8];
+        // ...
+        // For write ports, concat all the vector elements and assign to the
+        // write port. Example for a vector with 4 elements. Mem_writePort =
+        // {wrapper_writePort[3],wrapper_writePort[2],wrapper_writePort[1],wrapper_writePort[0]}
+        // Concat all mask bits to create a single multibit mask.
+        auto fVecType = fieldType.dyn_cast<FVectorType>();
+        auto elemWidth = fVecType.getElementType().getBitWidthOrSentinel();
+        WireOp lastIndexWire;
+        // Map vector elements to memory fields.
+        // Read data is mapped to vector elements using bit extraction.
+        // Write data is mapped from vector elements using bit concat.
+        // Mask data is mapped from vector elements using bit concat.
+        for (size_t vecI = 0, numElems = fVecType.getNumElements();
+             vecI < numElems; ++vecI) {
+          auto vI = builder.create<SubindexOp>(op.getLoc(), fieldValue, vecI);
+          if (isReadPort) {
```

Comment:
I'm wondering if inverting the order of the `if (isReadPort)` and `for` loop will result in simpler code, since it seems like very little is shared between the read vs. write cases besides the `for` loop itself.